src/libPMacc/include/particles/operations/updateCellListOfFrame.kernel

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright 2014-2017 Marco Garten, Alexander Matthes
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#pragma once
+
+#include "memory/shared/Allocate.hpp"
+#include "memory/Array.hpp"
+
+namespace PMacc
+{
+namespace particles
+{
+namespace operations
+{
+namespace kernel
+{
+
+struct UpdateCellListOfFrame
+{
+    template
+    <
+        size_t TileSize,
+        bool has_no_gaps = true,
+        typename T_CellEntryPoint,
+        typename T_CellCount,
+        typename T_FramePtr,
+        typename T_FrameNr
+    >
+    DINLINE void operator()(
+        T_CellEntryPoint& cellEntryPoint,
+        T_CellCount& cellCount,
+        T_FramePtr& frame,
+        T_FrameNr& frameNr
+    ) const
+    {
+        auto myParticle = frame[threadIdx.x];
+        if (has_no_gaps || myParticle[multiMask_])
+        {
+            //absolute position in frame list
+            auto myID = threadIdx.x + frameNr * TileSize;
+            auto & myCell = myParticle[localCellIdx_];
+            myParticle[inCellOffset_] = ::atomicExch(&(cellEntryPoint[myCell]),myID);
+            ::atomicAdd(&(cellCount[myCell]),1);
+        }
+        __syncthreads( );
+    }
+};
+
+struct UpdateCellListOfFramesKernel
+{
+    template<class ParBox, class Mapping>
+    DINLINE void operator()( ParBox pb, Mapping mapper ) const
+    {
+        using namespace particles::operations;
+
+        enum
+        {
+            TileSize = math::CT::volume<typename Mapping::SuperCellSize>::type::value,
+            Dim = Mapping::Dim
+        };
+
+        typedef typename ParBox::FramePtr FramePtr;
+
+        DataSpace<Dim> superCellIdx( mapper.getSuperCellIndex( DataSpace<Dim > (blockIdx) ) );
+
+        //data copied from right (last) to left (first)
```

Comment:
Well, I have now idea, what the comment meant, so I am afraid to touch it. Probably some secret language between René and Axel :smiling_imp: